### PR TITLE
fix: events loading inconsistency

### PR DESCRIPTION
**Pull request by gikf**
- [x] I have read [Chapter's contributing guidelines](https://github.com/freeCodeCamp/chapter/blob/main/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request targets the `main` branch of Chapter.

---
- Fixes double load on opening `/events`, one coming from the normal query and one from `fetchMore` after `currentPage` is changed/set.
- Fetching more is now limited to not visited pages numbers. Otherwise page with events is loaded from cache (I assume it's nextjs' cache). This might be a little overkill for now, both in doing so in first place and in using set to keep track of visited pages.

### CODE REVIEW

**Review by ojeytonwilliams — COMMENTED** [review]
Just to confirm: is the goal to prevent the client repeatedly requesting pages it has already fetched? e.g. if open events and click back and forth between pages 1 and 2 the browser should only make two GraphQL requests.

**Comment by gikf** [comment]
> Just to confirm: is the goal to prevent the client repeatedly requesting pages it has already fetched? e.g. if open events and click back and forth between pages 1 and 2 the browser should only make two GraphQL requests.

As an additional effect - yes. The main goal was to stop loading, fetching and rendering twice when `/events` is opened for the first time.

**Comment by ojeytonwilliams** [comment]
> As an additional effect - yes. The main goal was to stop loading, fetching and rendering twice when /events is opened for the first time.

I only see the one graphql request when I reload events.  Either way, I think this PR makes sense.

**Comment by gikf** [comment]
Well, this is awkward. Scratch the fetching then please.

I've double checked how to reproduce the weird behavior that made us all gather here:
- Open home page.
- Navigate to _Events_ from _Menu_.
- Navigate to home page.
- Navigate to _Events_ from _Menu_. (Additionally this step often gets stuck with _Loading..._, in such case circling back home page -> Events would get it unstuck).
- List with events are displayed (that would be cache I believe now), but then _Loading..._ appears, and fresh list with events is fetched.

I've also found that's not fully resolved. I'll investigate further, but fresh eyes could spot something obvious.

File: client/src/modules/events/pages/eventsPage.tsx [hunk 1]
```diff
@@ -54,17 +54,20 @@ function Pagination({
 const pageSize = 5;
 export const EventsPage: NextPage = () => {
   const [currentPage, setCurrentPage] = useState(1);
+  const [visitedPages, setVisitedPages] = useState(new Set([1]));
```
Comment (ojeytonwilliams) [inline client/src/modules/events/pages/eventsPage.tsx:57]:
```suggestion
  const [visitedPages, setVisitedPages] = useState(new Set([1]));
  const offset = (currentPage - 1) * pageSize;
```

File: client/src/modules/events/pages/eventsPage.tsx [hunk 2]
```diff
@@ -54,17 +54,20 @@ function Pagination({
...
   const { loading, error, data, fetchMore } = usePaginatedEventsWithTotalQuery({
     variables: { offset: (currentPage - 1) * pageSize, limit: pageSize },
```
Comment (ojeytonwilliams) [inline client/src/modules/events/pages/eventsPage.tsx:59]:
```suggestion
    variables: { offset, limit: pageSize },
```

File: client/src/modules/events/pages/eventsPage.tsx [hunk 3]
```diff
@@ -54,17 +54,20 @@ function Pagination({
...
   });
 
   useEffect(() => {
+    if (visitedPages.has(currentPage)) return;
     fetchMore({
       variables: {
         offset: data?.paginatedEventsWithTotal.events.length || 0,
```
Comment (ojeytonwilliams) [inline client/src/modules/events/pages/eventsPage.tsx:66]:
```suggestion
        offset,
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I like what you've done to reduce queries, but it got me thinking about why it was getting stuck on loading.  It seems like the incorrect offset was confusing it.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**